### ISSUE TITLE
Refactor rewrite logging

### ISSUE THREAD

**Issue by JohannesPeter1984** [querqy/querqy#375]
Currently, users of Querqy have different options to evaluate what is happening under the hood:
1. **Solr**
* Activating the debugging via debugQuery=on (rewriting action objects are are added as strings to the Solr output)
* Activating the info logging via querqy.infoLogging=on + sinks configured in solrconfig.xml (prints rewriter ids + number of the rule, but not the actual action)

2. **Elasticsearch**
* Activating the info logging via rewriter configuration (REST) with three different options (REWRITER_ID, DETAIL, NONE). While the basic rewriter information is covered as json, the details are contianed in a string.

As the information about rewriter activities is very critical for debugging, logging and tracking purposes, the capabilities of getting this information should be 
* consolidated via a common interface and a common response, as well as
* more API-ready, which means that details should be retrievable in a json format.

Object structure:

```json
"rewrite_logging": {
  "rewrite_chain": [
    {
      "$rewriter_id": [
        {
          "message": "iphone#243",
          "match": {
            "term": "iphone", 
            "type": "exact"
          },
          "action": {
            "input": "iphone",
            "instructions": [
              {
                "up": {
                  "param": "10.0",
                  "value": "apple smartphone"
                }
              },
              {
                "decorate": {
                  "value": ""
                }
              }
            ]
          },
          "properties": {
            "_id": "..."
          }
        }
      ],  
    }, 
    {
      "$rewriter_id": [
          "..."
      ]
    }
  ]
}

```

**Comment by JohannesPeter1984** [comment]
@mkr 
@renekrie

**Comment by JohannesPeter1984** [comment]
As this will require a model to create context objects (or data classes/objects) for input and instructions, we will have a code redundancy for a certain amount of time, as such classes already exist. However, the existing classes are not suitable for reflecting a context as they contain a lot of code / logic.

**Comment by renekrie** [comment]
Cross linking with #317 and Querqy for ES issue https://github.com/querqy/querqy-elasticsearch/issues/11

**Comment by JohannesPeter1984** [comment]
Furthermore, this implementation will include a breaking change with the current action printing in Solr (using debugQuery=on and the object contained in debug.query.rewrite. This might have an impact on Quepid @epugh.

**Comment by renekrie** [comment]
If we move the action printing to the infoLogging output, we'd need to enable the `response` infoLogging sink by default because otherwise users would have to turn on infoLogging before they could see this output for debugging

**Comment by JohannesPeter1984** [comment]
Related issues:

- https://github.com/querqy/querqy/issues/317
- https://github.com/querqy/querqy/issues/255
- https://github.com/querqy/querqy/issues/184
- https://github.com/querqy/querqy/issues/161

**Comment by JohannesPeter1984** [comment]
> If we move the action printing to the infoLogging output, we'd need to enable the `response` infoLogging sink by default because otherwise users would have to turn on infoLogging before they could see this output for debugging

I planned to include logging for every rewriter in the case of debug=on

**Comment by JohannesPeter1984** [comment]
After a few discussions about API designs, I changed the output design to the following structure:

```json
"rewrite_logging": {
  "rewrite_chain": [
    {
      "rewriter_id": "synonyms",
      "actions": [
        {
          "message": "iphone#243",
          "match": {
            "term": "iphone", 
            "type": "exact"
          },
          "action": {
            "input": "iphone",
            "instructions": [
              {
                "type": "up",
                "param": "10.0",
                "value": "apple smartphone"
              },
              {
                "type": "decorate",
                "value": ""
              }
            ]
          },
          "properties": {
            "_id": "..."
          }
        }
      ],  
    }, 
    {
      "$rewriter_id": [
          "..."
      ]
    }
  ]
}
```

**Comment by renekrie** [comment]
Implemented by #376, #387 and #400
